### ISSUE TITLE
HTTP2 pushes not working

### ISSUE THREAD

**Issue by vladbondarenko** [mholt/caddy#1573]
(Are you asking for help with using Caddy? Please use our forum instead: https://forum.caddyserver.com. If you are filing a bug report, please take a few minutes to carefully answer the following questions. If your issue is not a bug report, you do not need to use this template. Thanks!)

### 1. What version of Caddy are you running (`caddy -version`)?
Caddy 0.9.5 (+7d15435

### 2. What are you trying to do?
enable http2 push

### 3. What is your entire Caddyfile?
my.domain:443 {
push / {
/static/_cache/merged/d1da6bf60541e004556afd2b8358df2d.min.js
/static/_cache/merged/911bac75bca05a0baeedac6b4ebef29a.min.css
}
gzip
log /var/log/my.domain-access.log
errors /var/log/my.domain-error.log
proxy / 127.0.0.1:80 {
header_upstream Host "my.domain"
header_upstream X-Forwarded-Proto "https"
}
tls my.domain.crt my.domain.key
root /var/www/html
}

### 4. How did you run Caddy (give the full command and describe the execution environment)?
/usr/bin/caddy -quic -log stdout -agree=true -conf=/etc/caddy/Caddyfile -root=/var/www/html

### 5. Please paste any relevant HTTP request(s) here.

Checked Network on Chrome and FF.

### 6. What did you expect to see?
/static/_cache/merged/d1da6bf60541e004556afd2b8358df2d.min.js
/static/_cache/merged/911bac75bca05a0baeedac6b4ebef29a.min.css
to be pushed

### 7. What did you see instead (give full error messages and/or log)?
no Link headers with pushed resources while GETting / and them are not pushed.

Can you please check why pushes not working.
Thanks.

**Comment by mholt** [comment]
Hi Vlad, can you post a screenshot of your network tab for the requests of those resources? Make sure the Protocol and Type columns are visible.

As I understand it, Link headers are not set by the push middleware, but they are _read_ by the push middleware.

**Comment by vladbondarenko** [comment]
Ok, 

I`m using Caddy->Varnish->Nginx

I set 
add_header Link: </static/_cache/merged/d1da6bf60541e004556afd2b8358df2d.min.js>; rel=preload; as=script;
in nginx.

I set 
push / {
   method GET
   header x-http2-push 1
   /pub/static/_cache/merged/d1da6bf60541e004556afd2b8358df2d.min.js
   /pub/static/_cache/merged/911bac75bca05a0baeedac6b4ebef29a.min.css
}
in caddy to check if obj is pushed (if header presents)

I got  curl -SLIXGET https://mydomain
HTTP/1.1 200 OK
Accept-Ranges: bytes
Access-Control-Allow-Origin: *
Alt-Svc: quic=":443"; ma=2592000; v="36,35"
Alternate-Protocol: 443:quic
Cache-Control: max-age=0, must-revalidate, no-cache, no-store
Content-Type: text/html; charset=UTF-8
Date: Thu, 13 Apr 2017 04:03:22 GMT
Expires: Wed, 13 Apr 2016 04:03:21 GMT
Link: </static/_cache/merged/d1da6bf60541e004556afd2b8358df2d.min.js>; rel=preload; as=script;
Pragma: no-cache
Server: ESF
Set-Cookie: PHPSESSID=cqbjtvljb1gdltark5uvc8o2u3; expires=Thu, 13-Apr-2017 05:03:21 GMT; Max-Age=3600; path=/; domain=mydomain; HttpOnly
Set-Cookie: mage-messages=%5B%5D; expires=Fri, 13-Apr-2018 04:03:22 GMT; Max-Age=31536000; path=/
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Varnish: 2
X-Xss-Protection: 1; mode=block
Transfer-Encoding: chunked


Then i check Network in Chrome for pushed JS
Response headers are:
accept-ranges:bytes
access-control-allow-origin:*
alt-svc:quic=":443"; ma=2592000; v="36,35"
alternate-protocol:443:quic
cache-control:max-age=31536000, public
content-encoding:gzip
content-length:425526
content-type:application/javascript; charset=UTF-8
date:Thu, 13 Apr 2017 04:03:37 GMT
etag:W/"58de3a37-1c1ee7"
expires:Fri, 13 Apr 2018 04:03:37 GMT
last-modified:Fri, 31 Mar 2017 11:15:03 GMT
server:ESF
status:200
vary:Accept-Encoding
x-frame-options:SAMEORIGIN
x-varnish:32779 98308

but should be additional x-http2-push=1 header in resp of pushed object. 

I also use https://http2-push.io/ to check, but "No pushes found"

![screen16](https://cloud.githubusercontent.com/assets/9531461/24990074/dd11c038-2019-11e7-8bfe-f7e17ee70e79.png)
![screen15](https://cloud.githubusercontent.com/assets/9531461/24990075/dd153f6a-2019-11e7-9cd9-68467a957742.png)
![screen17](https://cloud.githubusercontent.com/assets/9531461/24990080/dd764698-2019-11e7-99a8-361c8fc7f082.png)

**Comment by mholt** [comment]
Hmm, what if you disable quic?

**Comment by vladbondarenko** [comment]
The same, except protocol. h2 instead of http2+quic35

**Comment by vladbondarenko** [comment]
I`m trying 

import (
        "net/http"
        "strings"
        "log"
        "github.com/mholt/caddy/caddyhttp/httpserver"
)

func (h Middleware) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
log.Printf("Enter push: %v")
        pusher, hasPusher := w.(http.Pusher)

in push/handler.go to see if any log will appear

but no "Enter push" in any log.

i dont know Go well but seems some functions in push handler not working.

in setup.go:
        httpserver.GetConfig(c).AddMiddleware(func(next httpserver.Handler) httpserver.Handler {
                log.Printf("Enter push3: %v")
                return Middleware{Next: next, Rules: rules}
        })

i got 
2017/04/14 08:43:04 Enter push3: %!v(MISSING)
2017/04/14 08:43:04 Enter push3: %!v(MISSING)
in /var/log/caddy.log
but not get 
log.Printf("Enter push: %v") from func (h Middleware) ServeHTTP

**Comment by mholt** [comment]
Oh, this is because the push handler is executed after proxy, in other words, the request is getting proxied before the push handler will get to see it, and it won't, because proxy terminates the chain.

@wendigo Do you know if the placement of push behind proxy was intentional? I can't recall...

**Comment by vladbondarenko** [comment]
Matt, what if trust app (backend behind proxy directive) and push resources only from response Link headers? (like nghttp2)

App should know what will be pushed and http server should trust.

Your thoughts?

**Comment by mholt** [comment]
> Matt, what if trust app (backend behind proxy directive) and push resources only from response Link headers? (like nghttp2)

The push middleware does this, the handler just doesn't appear to be ordered properly.

> I mean no need "push" directive at all.

We'll get there: but before we blanket-enable this (disabling things is often more tedious/tricky to do well) we need to work out a lot of details with push.

**Comment by vladbondarenko** [comment]
For now i put push before proxy. Pushes seems working but "header" in Caddyfile ignored. 
https://http2-push.io/ finaly found something.

**Comment by vladbondarenko** [comment]
![screen18](https://cloud.githubusercontent.com/assets/9531461/25062634/6398ddf2-21d8-11e7-8c21-561bc206e2cb.png)

Push work pretty cool, but 1 thing i see, initialy connection started using h2, and only after page reload (or loading static files) it starts quic.
Is that bug or feature? )

**Comment by mholt** [comment]
It looks like QUIC+h2push aren't operational yet in our (excellent) library of choice: https://github.com/lucas-clemente/quic-go/issues/372

Okay, so we need to reorder the push directive. Any objection, @wendigo? If not, or I don't hear by Tuesday evening or Wednesday, I'll probably make the change myself for release on Thursday. :+1:

**Comment by vladbondarenko** [comment]
Just 1 thing i forget to test.

If not add push resources to config, push module not automaticaly push them if Link resp header presents.
There is need to put those resources twice, in App and in Caddy as well.
but that is not a big issue i guess, push is working and thats cool.

**Comment by mholt** [comment]
@vladbondarenko Can you clarify? Does that mean you put `push` in your Caddyfile without any other arguments (i.e. on its own line) - is that what you tried, and it still doesn't push resources from the Link header?

**Comment by vladbondarenko** [comment]
if i put "push" at new line i get
2017/04/17 15:50:39 /etc/caddy/Caddyfile:8 - Parse error: Wrong argument count or unexpected line ending after 'push'

if i put "push /"
i get 2017/04/17 15:51:31 /etc/caddy/Caddyfile:8 - Parse error: Rule / has empty push resources list

Only works if i add args as resources. 

Should it work just with 
"push"
?

**Comment by mholt** [comment]
Ah, I didn't think about that. @wendigo What do you think? I think `push` by itself should be allowed, so that it can use the Link headers. (Also, the reordering of the directive -- I'm going to do that tomorrow unless I hear back that it's a bad idea.)

**Comment by mholt** [comment]
@vladbondarenko I've just pushed ce2a9cd which re-orders `push` to before `proxy` so the Link headers are read, and then I also made it so you can use `push` alone on a line just for pushing resources in the Link header.

Can you build the latest from master and see if that resolves this issue?

**Comment by vladbondarenko** [comment]
Yup, "push" alone and push re-order works now.

Thanks a lot.

**Comment by wendigo** [comment]
@mholt: sorry for not answering this - I'm in a process of changing my job - quite busy right now :( Thank you for fixing this. You change looks good! Can't wait for next version release :)

**Comment by mholt** [comment]
Ah, no problem @wendigo! Good luck on changing jobs. I know that can be stressful. I appreciate your contributions and look forward to when you can contribute again!

**Comment by ricardoboss** [comment]
Hi, I'm trying to use push too and it doesn't work. Should I open a new issue or report it here?

**Comment by mholt** [comment]
New issue please, don't reopen closed ones.
